### PR TITLE
chore: add helper gcp related functions for foundation deploy tasks

### PR DESCRIPTION
**Pull request by daniel-cit**
This PR adds helper gcp related functions for foundation deploy tasks

### CODE REVIEW

File: helpers/deployer/gcp/gcp.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+// NewGCP creates a new wrapper for Google Cloud Platform CLI.
+func NewGCP() GCP {
+	return GCP{
+		Runf:      gcloud.Runf,
+		sleepTime: 20,
+	}
+}
+
+// GetBuilds gets all Cloud Build builds form a project and region that satisfy the given filter.
+func (g GCP) GetBuilds(t testing.TB, projectID, region, filter string) []gjson.Result {
```
Comment (bharathkkb) [inline helpers/deployer/gcp/gcp.go:42]: nit: Would it be better to parse and return a map[id(string)]status(string) if that is what we need at the callsite?

File: helpers/deployer/gcp/gcp.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	return g.Runf(t, "builds list --project %s --region %s --limit 1 --sort-by ~createTime --filter %s", projectID, region, filter).Array()[0].Get("status").String()
+}
+
+// GetBuildStatus gets the status of the given build
+func (g GCP) GetBuildStatus(t testing.TB, projectID, region, buildID string) string {
+	return g.Runf(t, "builds describe %s  --project %s --region %s", buildID, projectID, region).Get("status").String()
+}
+
+// GetRunningBuildID gets the current build running for the given prohec, region, and filter
+func (g GCP) GetRunningBuildID(t testing.TB, projectID, region, filter string) string {
```
Comment (bharathkkb) [inline helpers/deployer/gcp/gcp.go:57]: Would it be better to receive a commit sha to filter by?

File: helpers/deployer/gcp/gcp.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+func (g GCP) GetLastBuildStatus(t testing.TB, projectID, region, filter string) string {
+	return g.Runf(t, "builds list --project %s --region %s --limit 1 --sort-by ~createTime --filter %s", projectID, region, filter).Array()[0].Get("status").String()
+}
+
+// GetBuildStatus gets the status of the given build
+func (g GCP) GetBuildStatus(t testing.TB, projectID, region, buildID string) string {
+	return g.Runf(t, "builds describe %s  --project %s --region %s", buildID, projectID, region).Get("status").String()
+}
+
+// GetRunningBuildID gets the current build running for the given prohec, region, and filter
```
Comment (bharathkkb) [inline helpers/deployer/gcp/gcp.go:56]:
```suggestion
// GetRunningBuildID gets the current build running for the given project, region, and filter
```

File: helpers/deployer/gcp/gcp.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+func (g GCP) GetRunningBuildID(t testing.TB, projectID, region, filter string) string {
+	time.Sleep(g.sleepTime * time.Second)
+	builds := g.GetBuilds(t, projectID, region, filter)
+	for _, build := range builds {
+		status := build.Get("status").String()
+		if status == "QUEUED" || status == "WORKING" {
```
Comment (bharathkkb) [inline helpers/deployer/gcp/gcp.go:62]: nit: let's move these to constants

File: helpers/deployer/gcp/gcp.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	return ""
+}
+
+// GetFinalBuildState gets the terminal status of the given build. It will wait if build is not finished.
+func (g GCP) GetFinalBuildState(t testing.TB, projectID, region, buildID string) string {
+	var status string
+	fmt.Printf("waiting for build %s execution.\n", buildID)
+	status = g.GetBuildStatus(t, projectID, region, buildID)
+	fmt.Printf("build status is %s\n", status)
+	for status != "SUCCESS" && status != "FAILURE" && status != "CANCELLED" {
```
Comment (bharathkkb) [inline helpers/deployer/gcp/gcp.go:75]: nit: might be good to have a max retry too so we exit

File: helpers/deployer/gcp/gcp.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+		fmt.Printf("build status is %s\n", status)
+		time.Sleep(g.sleepTime * time.Second)
+		status = g.GetBuildStatus(t, projectID, region, buildID)
+	}
+	fmt.Printf("final build status is %s\n", status)
+	return status
+}
+
+// WaitBuildSuccess waits for teh current build in a repo to finish.
```
Comment (bharathkkb) [inline helpers/deployer/gcp/gcp.go:84]:
```suggestion
// WaitBuildSuccess waits for the current build in a repo to finish.
```

File: helpers/deployer/gcp/gcp.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+type GCP struct {
+	Runf      func(t testing.TB, cmd string, args ...interface{}) gjson.Result
+	sleepTime time.Duration
+}
+
+// NewGCP creates a new wrapper for Google Cloud Platform CLI.
+func NewGCP() GCP {
+	return GCP{
+		Runf:      gcloud.Runf,
```
Comment (bharathkkb) [inline helpers/deployer/gcp/gcp.go:36]: great idea to stub these!

**Review by bharathkkb — APPROVED** [review]

**Comment by shark696969** [comment]
Is:received solved all

Pada tanggal Sel, 4 Apr 2023 07.25, Bharath KKB ***@***.***>
menulis:

File: helpers/deployer/gcp/gcp.go (hunk 2 above)
Comment (daniel-cit) [reply helpers/deployer/gcp/gcp.go:73]: done

File: helpers/deployer/gcp/gcp.go (hunk 3 above)
Comment (daniel-cit) [reply helpers/deployer/gcp/gcp.go]: fixed

File: helpers/deployer/gcp/gcp.go (hunk 5 above)
Comment (daniel-cit) [reply helpers/deployer/gcp/gcp.go]: fixed

File: helpers/deployer/gcp/gcp.go (hunk 6 above)
Comment (daniel-cit) [reply helpers/deployer/gcp/gcp.go]: fixed

File: helpers/deployer/gcp/gcp.go (hunk 4 above)
Comment (daniel-cit) [reply helpers/deployer/gcp/gcp.go]: done

File: helpers/deployer/gcp/gcp.go (hunk 1 above)
Comment (daniel-cit) [reply helpers/deployer/gcp/gcp.go]: done

**Review by bharathkkb — APPROVED** [review]

**Pull request merged**